Sort interface materials

lindsayad opened the pull request:
closes #17171

@arovinelli can you see if this resolves your issue?

lindsayad committed:
Sort interface materials

closes #17171

lindsayad commented:
Hmm this doesn't appear to be working on a simple test I drew up...

arovinelli commented:
@lindsayad for some reason it works on my case. Now the execution order is correct

lindsayad commented:
Yea I realized what the problem was. I am using AD material properties and we didn't have the correct methods for AD material properties in `InterfacMaterial` yet

lindsayad commented:
Thanks for reporting this issue!

arovinelli commented:
> Thanks for reporting this issue!

You are welcome, in the next few days I'll also try with stateful properties. If anything comes up I'll let you know :)

lindsayad committed:
Implement AD material property getters in InterfaceMaterial

We must ensure that these methods exist otherwise the base
`MaterialPropertyInterface` methods will be used and then we won't
register requested material properties for dependency sorting

lindsayad committed:
Test interface material sorting

GiudGiud commented on framework/doc/content/syntax/Materials/index.md:
```diff
@@ -197,6 +197,14 @@ table lists the types of properties that are available for automatic output.
+Materials are sorted such that one material may consume a property produced by
+another material and know that the consumed property will be up-to-date,
+e.g. the producer material will execute before the consumer material. If a
+cyclic dependcy is detected between two materials, then MOOSE will produce an
```
```suggestion
cyclic dependency is detected between two materials, then MOOSE will produce an
```

GiudGiud reviewed (commented):
lgtm

lindsayad committed:
Update framework/doc/content/syntax/Materials/index.md

Co-authored-by: Guillaume Giudicelli <guillaume.giudicelli@gmail.com>

lindsayad commented:
@rwcarlsen could you re-approve after I applied @GiudGiud 's suggestion?
